Update containerize to prep for shipping

BenVillalobos opened the pull request:
### Quick Summary
The biggest change is creating a CreateNewImageToolTask who's job it is to call `dotnet containerize.dll --all-the-args`. It There's not a branching target, `PublishContainer` is defined differently based on `MSBuildRuntimeType`.

**NOTE:** Skip over the `Containerize` changes, the majority of those changes are happening in https://github.com/dotnet/sdk-container-builds/pull/155

baronfel commented on containerize/Program.cs:
```diff
@@ -43,7 +43,10 @@
-    description: "The tags to associate with the new image.");
+    description: "The tags to associate with the new image.")
+{
+    AllowMultipleArgumentsPerToken = true
```
Similar for the other 'multiple tokens' settings below. I don't really have an opinion on it.

baronfel commented on packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -95,4 +97,21 @@
+    <Target Name="PublishContainer" DependsOnTargets="$(PublishContainerDependsOn)"
+                                    Condition="'$(MSBuildRuntimeType)' == 'Full'">
+
+        <CreateNewImageToolTask ToolDirectory="$(ContainerizeFolder)"
```
Is this line where we'd want to use `NetCoreRoot` like we saw in the other tool tasks?

BenVillalobos commented on Microsoft.NET.Build.Containers/ContainerHelpers.cs:
```diff
@@ -96,9 +96,15 @@ public static bool IsValidImageTag(string imageTag)
+                                                            out string? containerRegistry,
+                                                            out string? containerName,
+                                                            out string? containerTag)
+#endif
```
I swear I did this yesterday and it didn't work, Today, it does 🤷‍♂️

BenVillalobos commented:
cc @baronfel https://github.com/dotnet/sdk-container-builds/pull/164/commits/8d502676d5587402c233319e6ae01052f6a24582 here's where I merged your stuff into mine.

Note: I re-defined some of your stuff (PortType and Port) and moved it into a whole new file. That file (net472Definitions.cs) is only added in the `net472` build and removed in the `net7.0` build. It's...not ideal, but importing `Image.cs` into `net472` was a whole other yak to be shaved that I didn't want to jump into.

The new `IsExternalInit` is a workaround to allow a `record` type.

BenVillalobos commented:
Just tested again locally to extra-verify: 👍

baronfel commented on Microsoft.NET.Build.Containers/Microsoft.NET.Build.Containers.csproj:
```diff
@@ -27,6 +27,14 @@
   <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
     <Compile Remove="*.*" />
     <!-- A separate PR will add the required CreateNewImageToolTask file. -->
```
```suggestion
```

baronfel commented on Microsoft.NET.Build.Containers/ParseContainerProperties.cs:
```diff
@@ -161,8 +164,8 @@ public override bool Execute()
             {
-                Log.LogWarning(null, "CONTAINER001", "Container.InvalidImageName", null, 0, 0, 0, 0, $"{nameof(ContainerImageName)} was not a valid container image name, it was normalized to {normalizedImageName}");
-                NewContainerImageName = normalizedImageName;
+                Log.LogWarning(null, "CONTAINER001", "Container.InvalidImageName", null, 0, 0, 0, 0, $"{nameof(ContainerImageName)}:'{ContainerImageName}' was not a valid container image name, it was normalized to {normalizedImageName}");
```
nice addition here 👍

baronfel commented on packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -34,7 +34,7 @@
             <ContainerRegistry Condition="'$(ContainerRegistry)' == ''">docker://</ContainerRegistry>
             <!-- Note: spaces will be replaced with '-' in ContainerImageName and ContainerImageTag -->
-            <ContainerImageName Condition="'$(ContainerImageName)' == ''">$(AssemblyName)</ContainerImageName>
+            <ContainerImageName Condition="'$(ContainerImageName)' == ''">$(AssemblyName.ToLowerInvariant())</ContainerImageName>
```
this shouldn't need to be done in the targets, we do the normalization in [this method](https://github.com/dotnet/sdk-container-builds/blob/tim-the-tooltask-taylor/Microsoft.NET.Build.Containers/ContainerHelpers.cs#L148-L166) - was there a reason that wasn't working?

BenVillalobos replied:
So the problem (from my perspective) was that running `msbuild /t:Publish /p:RuntimeIdentifier=linux-x64 /p:Configuration=Release /p:PublishProfile=DefaultContainer` is essentially saying "do all the defaults", but the default led to a capitalized assemblyname, which led to a warning in the "default" case.

Thoughts on this? Is the warning okay here?

BenVillalobos replied:
We decided to remove this from the targets, let the task normalize it and log a high-importance message instead of warning.

baronfel replied:
Justification for this is that the task/targets should be _very_ focused on doing the right thing by default. Telling a user to do a thing when we could have just done it feels bad, especially when the response to us saying 'hey you can't have these characters' is going to be writing the same regex and replacements that we already do.

rainersigwald replied:
I would ideally like to have a warning _if the user specified a bad image name_, but warningless correction _if we inferred the name ourselves_. Think that's worth going to pains for?

baronfel replied:
This would require some type of flag set when we did inference that we'd pass into the tasks? so during the inference phase we'd set `<_ContainerDidInferImageName>true</_ContainerDidInferImageName>`, flow that through as a boolean input to both tasks, and then pass it into the validation phase to let us know whether to warn or not?

baronfel reviewed (approved):
Looks really good! I left a couple notes, just areas where it wasn't immediately clear to me.

rainersigwald commented on packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -96,4 +98,29 @@
+            <DotNetHostFileName Condition="'$(OS)' == 'Windows_NT'">dotnet.exe</DotNetHostFileName  >
+        </PropertyGroup>
+
+        <CreateNewImageToolTask ContainerizeDirectory="$(ContainerizeFolder)"
```
Looks like this is missing `ExposedPorts`, which I think is just a symptom of my main concern about this PR which is: why do we need two separate task invocations?

baronfel replied:
The new task also includes the DotNetRoot location parts, that might need some massaging to 'just work' inside the same Task invocation

BenVillalobos replied:
So are we thinking we only have `CreateNewImageToolTask` or (CreateNewImage becomes a tooltask itself) and we check inside the task itself if we're on Core/Full?

BenVillalobos replied:
Cleared this up. The ToolTask is being renamed to CreateNewImage

rainersigwald commented on packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -96,4 +98,29 @@
     </Target>
+
+    <Target Name="PublishContainer" DependsOnTargets="$(PublishContainerDependsOn)"
+                                    Condition="'$(MSBuildRuntimeType)' == 'Full'">
```
Unfortunately, conditions on tasks don't work this way--they're evaluated just before the task runs, so this target always wins and just won't run on Core, ever.

```
Target Name=PublishContainer Project=container-package-test.csproj
    Target "PublishContainer" skipped, due to false condition; ('$(MSBuildRuntimeType)' == 'Full') was evaluated as ('Core' == 'Full').
```

BenVillalobos replied:
Why would this target always win? Is this the only one defined in that case?

BenVillalobos commented:
![image](https://user-images.githubusercontent.com/4691428/190712080-cf87ea60-8275-4c6c-8f06-2e9ff9707c14.png)

things are looking good!

baronfel commented:
One thing to check on the tooltask - do the console outputs match the expected warning/error format that the MSBuild diagnostic regex is looking for?

BenVillalobos commented:
Something funky is going on, my test scenarios lock up now.  it might just be my machine but it's slowing me down.

BenVillalobos commented:
I'm running into quoting issues now: 
```
Directory does not exist: 'bin\Release\net7.0\linux-x64\publish"  --baseregistry https://mcr.microsoft.com --baseimagename dotnet/runtime-deps --baseimagetag 7.0 --outputregistry docker:// --imagename createnewimagetest --workingdirectory /app --entrypoint /app/CreateNewImageTest --labels org.opencontainers.image.created=2022-09-16T20:16:19.5478811Z --imagetags 1.0.0 --ports 80/tcp'.
```

There's a single and double quote in the directory here, I'm not sure where the path is getting eaten up.

BenVillalobos commented:
If works if I don't quote the `PublishDirectory`, I wonder if it has to do with that directory ending in `\`

BenVillalobos commented:
Should be good to go 👍

BenVillalobos commented:
@rainersigwald ping on requested changes. If all is well please squash👍

rainersigwald commented on Microsoft.NET.Build.Containers/ContainerBuilder.cs:
```diff
@@ -0,0 +1,76 @@
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Containerize: error CONTAINER001: Failed to push to output registry: {e}");
```
`Console.WriteLine` feels wrong: doesn't this sort of thing need to be written to the MSBuild log when this is called inproc?

baronfel replied:
This method isn't called from the in-proc pathway, only the console app. The thought here was to write in a format that the MSBuild error pattern would recognize (which I think we did get). I'm testing the display in VS this morning.

rainersigwald commented on Microsoft.NET.Build.Containers/CreateNewImageToolTask.cs:
```diff
@@ -0,0 +1,153 @@
+    /// The path to the folder containing `containerize.dll`.
+    /// </summary>
+    [Required]
+    public string ContainerizeDirectory { get; set; }
```
Nit for later refactor because it just occurred to me: make this + the other task a partial class, and pull all these input /output properties into a single file, to reduce maintenance.

baronfel replied:
oh of course. F# doesn't have partial classes so they never jump out to me as a design possibility.

BenVillalobos replied:
https://github.com/dotnet/sdk-container-builds/issues/171

rainersigwald commented on Microsoft.NET.Build.Containers/CreateNewImageToolTask.cs:
```diff
@@ -0,0 +1,153 @@
+
+    protected override string GenerateCommandLineCommands()
+    {
+        return Quote(ContainerizeDirectory + "containerize.dll") + " " +
```
For later refactor: this should use `Microsoft.Build.Utilities.CommandLineBuilder` (and @baronfel, we should probably expend some effort to make sure `CommandLineBuilder` plays well with S.CL).

BenVillalobos replied:
https://github.com/dotnet/sdk-container-builds/issues/172

BenVillalobos commented on Microsoft.NET.Build.Containers/ContainerHelpers.cs:
```diff
@@ -13,7 +13,28 @@ public static class ContainerHelpers
+    private static Regex imageNameCharacters = new Regex(@"[^a-z0-9_\-/]");
+
+#if !NET7_0_OR_GREATER
+    // Taken straight from dotnet/runtime for net472 visibility
```
Not sure how you saw this, it's out of the file.

BenVillalobos commented on Microsoft.NET.Build.Containers/net472Definitions.cs:
```diff
@@ -0,0 +1,36 @@
+        public bool ReturnValue { get; }
+    }
+
+    public enum PortType
+    {
+        tcp,
+        udp
+    }
+
+    public record Port(int number, PortType type);
```
This is to avoid a clash between files. Port is defined in `Image.cs` and importing that file into the net472 build would cause all sorts of other importing/net7.0 issues. I'll create a separate `ContainerTypes` class to house this stuff, and keep the `IsExternalInit` logic (to work around the net472 record type issue)/

BenVillalobos replied:
Update: I just placed it in Containerhelpers to reduce clutter

BenVillalobos commented on packaging/build/Microsoft.NET.Build.Containers.props:
```diff
@@ -1,14 +1,19 @@
         <!--The folder where the custom task will be present. It points to inside the nuget package. -->
-        <CustomTasksFolder>$(MSBuildThisFileDirectory)..\$(taskFoldername)\$(taskFramework)</CustomTasksFolder>
+        <CustomTasksFolder>$(MSBuildThisFileDirectory)..\$(taskFoldername)\$(taskFramework)\</CustomTasksFolder>
+        <ContainerizeFolder>$(MSBuildThisFileDirectory)..\$(ContainerizeFolderName)\</ContainerizeFolder>
```
https://github.com/dotnet/sdk-container-builds/issues/170
